crates/astria-composer/src/searcher/executor.rs

```diff
@@ -0,0 +1,321 @@
+    // sequencer.
+    sequencer_client: SequencerClient,
+    // Private key used to sign sequencer transactions
+    sequencer_key: SigningKey,
+    // Nonce of the sequencer account we sign with
+    nonce: Arc<AtomicU32>,
+    // The sequencer address associated with the private key
+    address: Address,
+}
+
+#[derive(Debug)]
+pub(super) struct Status {
+    is_connected: bool,
+}
+
+impl Status {
+    pub(super) fn new() -> Self {
+        Self {
+            is_connected: false,
+        }
+    }
+
+    pub(super) fn is_connected(&self) -> bool {
+        self.is_connected
+    }
+}
+
+/// The `Executor` interfaces with the sequencer. It handles account nonces, transaction signing,
+/// and transaction submission.
+/// The `Executor` receives `Vec<Action>` from the bundling logic, packages them with a nonce into
+/// an `Unsigned`, then signs them with the sequencer key and submits to the sequencer.
+/// Its `status` field indicates that connection to the sequencer node has been established.
+impl Executor {
+    pub(super) fn new(
+        sequencer_url: &str,
+        private_key: &SecretString,
+        executor_rx: mpsc::Receiver<Vec<Action>>,
+    ) -> eyre::Result<Self> {
+        // connect to sequencer node
+        let sequencer_client =
+            SequencerClient::new(sequencer_url).wrap_err("failed constructing sequencer client")?;
+
+        // create signing key for sequencer txs
+        let mut private_key_bytes: [u8; 32] = hex::decode(private_key.expose_secret())
+            .wrap_err("failed to decode private key bytes from hex string")?
+            .try_into()
+            .map_err(|_| eyre!("invalid private key length; must be 32 bytes"))?;
+        let sequencer_key =
+            SigningKey::try_from(private_key_bytes).wrap_err("failed to parse sequencer key")?;
+        private_key_bytes.zeroize();
+
+        // create address from signing key
+        let sequencer_address = Address::from_verification_key(sequencer_key.verification_key());
+
+        // create channel for status reporting
+        let (status, _) = watch::channel(Status::new());
+
+        Ok(Self {
+            status,
+            executor_rx,
+            sequencer_client,
+            sequencer_key,
+            nonce: Arc::new(AtomicU32::new(0)),
+            address: sequencer_address,
+        })
+    }
+
+    /// Return a reader to the status reporting channel
+    pub(super) fn subscribe(&self) -> watch::Receiver<Status> {
+        self.status.subscribe()
+    }
+
+    /// Return the current nonce
+    fn nonce(&self) -> u32 {
+        self.nonce.load(Ordering::Relaxed)
+    }
+
+    /// Gets the next nonce to sign over
+    fn get_next_nonce(&mut self) -> u32 {
+        // get current nonce and calculate next one
+        let curr_nonce = self.nonce();
+        let next_nonce = curr_nonce + 1;
+        // save next nonce
+        self.nonce.store(next_nonce, Ordering::Relaxed);
+        curr_nonce
```

Comment:
replace this with `fetch_add`: https://doc.rust-lang.org/stable/std/sync/atomic/struct.AtomicU32.html#method.fetch_add

```suggestion
self.nonce.fetch_add(1, Ordering::Relaxed)
```